src/main/java/frc/robot/utils/StateMachine.java

```diff
@@ -0,0 +1,190 @@
+import frc.robot.subsystems.NoteDetection;
+import frc.robot.subsystems.ShooterSubsystem;
+import frc.robot.subsystems.TransitSubsystem;
+
+public class StateMachine {
+    public enum State {
+        /**
+         * Close enough to a note to pickup
+         */
+        TARGETING_NOTE,
+        /**
+         * A note is in the intake but not in the transit
+         */
+        OWNS_NOTE,
+        /**
+         * A note is in the transit & ready to be shot
+         */
+        NOTE_LOADED,
+        /**
+         * Note is loaded and wheels are up to speed
+         */
+        READY_TO_SHOOT,
+        /**
+         * Robot does have a note and is not near a note
+         */
+        NO_NOTE
+    }
+
+    private State currentState = State.NO_NOTE;
+
+    private ShooterSubsystem m_shooterSubystem;
+    private ElevatorSubsystem m_elevatorSubsystem;
+    private TransitSubsystem m_transitSubsystem;
+    private ClimberSubsystem m_climberSubsystem;
+    private IntakeSubsystem m_intakeSubsystem;
+    private NoteDetection m_noteDetection;
+    private DriveSubsystem m_driveSubsystem;
+
+    private Debouncer m_noteDetectionDebouncer = new Debouncer(DriveConstants.kStatemachineDebounce.get(),
+            DebounceType.kBoth);
+    private Debouncer m_intakeLimitDebouncer = new Debouncer(DriveConstants.kStatemachineDebounce.get(),
+            DebounceType.kBoth);
+    private Debouncer m_transitLimitDebouncer = new Debouncer(DriveConstants.kStatemachineDebounce.get(),
+            DebounceType.kBoth);
+    private Debouncer m_shooterSpeedDebouncer = new Debouncer(DriveConstants.kStatemachineDebounce.get(),
+            DebounceType.kBoth);
+    private Debouncer m_elevatorAngleDebouncer = new Debouncer(DriveConstants.kStatemachineDebounce.get(),
+            DebounceType.kBoth);
+
+    private Command currentDesiredCommand = getCommandForState(currentState);
+
+    public StateMachine(ShooterSubsystem shooterSubsystem, ElevatorSubsystem elevatorSubsystem,
+            TransitSubsystem transitSubsystem, ClimberSubsystem climberSubsystem, IntakeSubsystem intakeSubsystem,
+            NoteDetection noteDetection, DriveSubsystem driveSubsystem) {
+        m_shooterSubystem = shooterSubsystem;
+        m_elevatorSubsystem = elevatorSubsystem;
+        m_transitSubsystem = transitSubsystem;
+        m_climberSubsystem = climberSubsystem;
+        m_intakeSubsystem = intakeSubsystem;
+        m_noteDetection = noteDetection;
+        m_driveSubsystem = driveSubsystem;
+    }
+
+    public void periodic() {
+        switch (currentState) {
+            case NO_NOTE:
+                // TODO: This will likely need to change when alex refactors the note detector
+                if (m_noteDetectionDebouncer.calculate(m_noteDetection.hasNote)) {
+                    currentState = State.TARGETING_NOTE;
+                    currentDesiredCommand = getCommandForState(currentState);
+                }
+                break;
+            case TARGETING_NOTE:
```

Comment:
What should happen if we lose track of the note we are targeting?

Reply:
That whole conditional will probably need to change, because we need to check if the note is within a certain distance. If we lose that note it should probably reset to the no note state, but I'm going to wait to do all that until alex finishes refactoring the note detector.